File: src/main/java/me/andre111/d20server/template/TemplateComponentPlaceholder.java
package me.andre111.d20server.template;

import java.util.ArrayList;
import java.util.List;

import me.andre111.d20common.model.chat.ChatEntry.TriggeredContent;
import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20common.scripting.ScriptException;
import me.andre111.d20common.scripting.expression.DiceRoll;

public final class TemplateComponentPlaceholder extends TemplateComponent {
	private final int index;
	private final Placeholder placeholder;
	private String string = null;
	private List<DiceRoll> diceRolls = null;
	private List<TriggeredContent> triggeredContent = null;
	
	public TemplateComponentPlaceholder(int index, Placeholder placeholder) {
		this.index = index;
		this.placeholder = placeholder;
	}
	
	public int getIndex() {
		return index;
	}
	
	public void parse(Profile profile, String input) throws ScriptException {
		diceRolls = new ArrayList<>();
		triggeredContent = new ArrayList<>();
		string = placeholder.parse(profile, input, diceRolls, triggeredContent);
	}
	
	@Override
	public String getString() {
		if(string == null) {
			throw new IllegalStateException("Placeholder was not parsed before getting string.");
		}
		String currentString = string;
		string = null;
		return currentString;
	}

	@Override
	public List<DiceRoll> getDiceRolls() {
		if(diceRolls == null) {
			throw new IllegalStateException("Placeholder was not parsed before getting dice rolls.");
		}
		var currentDiceRolls = diceRolls;
		diceRolls = null;
		return currentDiceRolls;
	}

	@Override
	public List<TriggeredContent> getTriggeredContent() {
		if(triggeredContent == null) {
			throw new IllegalStateException("Placeholder was not parsed before getting triggered content.");
		}
		var currentTriggeredContent = triggeredContent;
		triggeredContent = null;
		return currentTriggeredContent;
	}
}

File: core/server/template/template-component-placeholder.js
import { TemplateComponent } from './template-component.js';

export class TemplateComponentPlaceholder extends TemplateComponent {
    index;
    placeholder;

    string = null;
    diceRolls = null;
    triggeredContent = null;

    constructor(index, placeholder) {
        super();

        this.index = index;
        this.placeholder = placeholder;
    }

    getIndex() {
        return this.index;
    }

    parse(profile, input) {
        this.diceRolls = [];
        this.triggeredContent = [];
        this.string = this.placeholder.parse(profile, input, this.diceRolls, this.triggeredContent);
    }

    getString() {
        if (this.string == null) throw new Error('Placeholder was not parsed before getting string');
        const currentString = this.string;
        this.string = null;
        return currentString;
    }

    getDiceRolls() {
        if (this.diceRolls == null) throw new Error('Placeholder was not parsed before getting dice rolls');
        const currentDiceRolls = this.diceRolls;
        this.diceRolls = null;
        return currentDiceRolls;
    }

    getTriggeredContent() {
        if (this.triggeredContent == null) throw new Error('Placeholder was not parsed before getting triggered content');
        const currentTriggeredContent = this.triggeredContent;
        this.triggeredContent = null;
        return currentTriggeredContent;
    }
}
